### PR TITLE
KAFKA-4655: Improve test coverage of CompositeReadOnlySessionStore

### PR DESCRIPTION
**Pull request by jeyhunkarimov**

### CODE REVIEW

**Comment by jeyhunkarimov** [comment]
call for review, @mjsax @dguy

File: streams/src/test/java/org/apache/kafka/streams/state/internals/CompositeReadOnlySessionStoreTest.java [hunk 1]
```diff
@@ -111,10 +112,35 @@ public void shouldThrowInvalidStateStoreExceptionOnRebalance() throws Exception
...
     }
 
-    @Test(expected = InvalidStateStoreException.class)
-    public void shouldThrowInvalidStateStoreExceptionIfFetchThrows() throws Exception {
-        underlyingSessionStore.setOpen(false);
-        underlyingSessionStore.fetch("key");
+    @Test
+    public void shouldThrowInvalidStateStoreExceptionIfFetchThrows() {
+        try {
+            underlyingSessionStore.setOpen(false);
```
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/state/internals/CompositeReadOnlySessionStoreTest.java]: It's better to have this outside the try block

File: streams/src/test/java/org/apache/kafka/streams/state/internals/CompositeReadOnlySessionStoreTest.java [hunk 2]
```diff
@@ -111,10 +112,35 @@ public void shouldThrowInvalidStateStoreExceptionOnRebalance() throws Exception
         store.fetch("a");
     }
 
-    @Test(expected = InvalidStateStoreException.class)
-    public void shouldThrowInvalidStateStoreExceptionIfFetchThrows() throws Exception {
-        underlyingSessionStore.setOpen(false);
-        underlyingSessionStore.fetch("key");
+    @Test
+    public void shouldThrowInvalidStateStoreExceptionIfFetchThrows() {
```
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/state/internals/CompositeReadOnlySessionStoreTest.java]: This should be two tests, IMHO.

File: streams/src/test/java/org/apache/kafka/streams/state/internals/CompositeReadOnlySessionStoreTest.java [hunk 3]
```diff
@@ -111,10 +112,35 @@ public void shouldThrowInvalidStateStoreExceptionOnRebalance() throws Exception
...
+        } catch (InvalidStateStoreException e) { }
+
+        try {
+            sessionStore.fetch("key");
+            fail("Should have thrown InvalidStateStoreException with session store");
+        } catch (InvalidStateStoreException e) { }
     }
 
+    @Test
+    public void shouldThrowNullPointerException() {
```
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/state/internals/CompositeReadOnlySessionStoreTest.java]: This should be three tests IMHO

**Review by mjsax — COMMENTED** [review]
Thanks a lot for writing tests!!!

File: streams/src/test/java/org/apache/kafka/streams/state/internals/CompositeReadOnlySessionStoreTest.java [hunk 4]
```diff
@@ -113,34 +113,35 @@ public void shouldThrowInvalidStateStoreExceptionOnRebalance() throws Exception
...
+        sessionStore.fetch(null);
+    }
 
-        try {
-            underlyingSessionStore.put(new Windowed<>((String) null, new SessionWindow(0, 0)), 1L);
-            fail("Should have thrown NullPointerException while putting null window key");
-        } catch (NullPointerException e) { }
+    @Test(expected = NullPointerException.class)
+    public void shouldThrowNullPointerExceptionIfPuttingNullSessionWindowKey() {
+        underlyingSessionStore.put(new Windowed<>((String) null, new SessionWindow(0, 0)), 1L);
```
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/state/internals/CompositeReadOnlySessionStoreTest.java]: It's unclear here, that `put` throws and not `new Windowed()` or `new SeeesionWindow()` -- should split and use try-fail-catch pattern here.

**Review by mjsax — CHANGES_REQUESTED** [review]
Thanks for the update. One more comment :)

**Review by mjsax — APPROVED** [review]
LGTM.

Call for final review and merging @guozhangwang

File: streams/src/test/java/org/apache/kafka/streams/state/internals/CompositeReadOnlySessionStoreTest.java [hunk 5]
```diff
@@ -111,10 +112,40 @@ public void shouldThrowInvalidStateStoreExceptionOnRebalance() throws Exception
...
+            fail("Should have thrown InvalidStateStoreException with underlying session store");
+        } catch (InvalidStateStoreException e) { }
+    }
+
+    @Test
+    public void shouldThrowInvalidStateStoreExceptionIfSessionFetchThrows() {
         underlyingSessionStore.setOpen(false);
-        underlyingSessionStore.fetch("key");
+        try {
+            sessionStore.fetch("key");
```
Comment (dguy) [inline streams/src/test/java/org/apache/kafka/streams/state/internals/CompositeReadOnlySessionStoreTest.java:119]: We should also add `Objects.requireNonNull(..)` in `CachingSessionStore.fetch(..)` as the NPE is not thrown directly by the code under test

File: streams/src/test/java/org/apache/kafka/streams/state/internals/CompositeReadOnlySessionStoreTest.java [hunk 6]
```diff
@@ -111,10 +112,40 @@ public void shouldThrowInvalidStateStoreExceptionOnRebalance() throws Exception
...
+    @Test(expected = NullPointerException.class)
+    public void shouldThrowNullPointerExceptionIfFetchingNullKey() {
+        sessionStore.fetch(null);
+    }
+
+    @Test
+    public void shouldThrowNullPointerExceptionIfPuttingNullSessionWindowKey() {
+        Windowed<String> sessionKey = new Windowed<>(null, new SessionWindow(0, 0));
+        try {
+            underlyingSessionStore.put(sessionKey, 1L);
```
Comment (dguy) [inline streams/src/test/java/org/apache/kafka/streams/state/internals/CompositeReadOnlySessionStoreTest.java]: should be: `sessionStore.put(...)`

File: streams/src/test/java/org/apache/kafka/streams/state/internals/CompositeReadOnlySessionStoreTest.java [hunk 7]
```diff
@@ -111,10 +112,40 @@ public void shouldThrowInvalidStateStoreExceptionOnRebalance() throws Exception
...
+            fail("Should have thrown NullPointerException while setting windowed sessionKey to null");
+        } catch (NullPointerException e) { }
+    }
+
+    @Test(expected = NullPointerException.class)
+    public void shouldThrowNullPointerExceptionIfPuttingNullKey() {
+        underlyingSessionStore.put(null, 1L);
```
Comment (dguy) [inline streams/src/test/java/org/apache/kafka/streams/state/internals/CompositeReadOnlySessionStoreTest.java]: `sessionStore.put(..)`  - should also add a null check in `CachingSessionStore.put(..)`

File: streams/src/test/java/org/apache/kafka/streams/state/internals/CompositeReadOnlySessionStoreTest.java [hunk 8]
```diff
@@ -111,10 +112,40 @@ public void shouldThrowInvalidStateStoreExceptionOnRebalance() throws Exception
         store.fetch("a");
     }
 
-    @Test(expected = InvalidStateStoreException.class)
-    public void shouldThrowInvalidStateStoreExceptionIfFetchThrows() throws Exception {
+    @Test
+    public void shouldThrowInvalidStateStoreExceptionIfUnderlyingSessionStoreFetchThrows() {
+        underlyingSessionStore.setOpen(false);
+        try {
+            underlyingSessionStore.fetch("key");
```
Comment (dguy) [inline streams/src/test/java/org/apache/kafka/streams/state/internals/CompositeReadOnlySessionStoreTest.java]: `sessionStore.fetch(...)`

**Comment by jeyhunkarimov** [comment]
@dguy I removed tests which involved `underlyingSessionStore` (`ReadOnlySessionStoreStub`). You suggested to change from 

`underlyingSessionStore.put(sessionKey, 1L);
`
to 

`sessionStore.put(sessionKey, 1L);
`
However, `sessionStore` does not have `put()` method. Moreover, testing `ReadOnlySessionStore` in `CompositeReadOnlySessionStoreTest` does not make sense I think.

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/5558/
Test PASSed (JDK 8 and Scala 2.12).

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/5574/
Test PASSed (JDK 7 and Scala 2.11).

File: streams/src/main/java/org/apache/kafka/streams/state/internals/CachingSessionStore.java [hunk 9]
```diff
@@ -150,11 +151,14 @@ public void put(final Windowed<K> key, AGG value) {
 
     @Override
     public KeyValueIterator<Windowed<K>, AGG> fetch(final K key) {
+        Objects.requireNonNull(key, "key cannot be null");
         return findSessions(key, 0, Long.MAX_VALUE);
     }
 
     @Override
     public KeyValueIterator<Windowed<K>, AGG> fetch(K from, K to) {
+        Objects.requireNonNull(from, "<from> cannot be null");
```
Comment (dguy) [inline streams/src/main/java/org/apache/kafka/streams/state/internals/CachingSessionStore.java]: nit: remove the `<` and `>` here and next line. Just to keep it consistent with everywhere else we do this

**Review by dguy — COMMENTED** [review]
One more very minor comment

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/5597/
Test PASSed (JDK 8 and Scala 2.12).

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/5611/
Test PASSed (JDK 7 and Scala 2.11).

**Review by dguy — APPROVED** [review]
Thanks @jeyhunkarimov LGTM

**Pull request closed without merging**